### ISSUE TITLE
Unable to load e_sqlite3 in Windows 8.1 project

### ISSUE THREAD

**Issue by Myndale** [praeclarum/sqlite-net#515]
I'm getting an error trying to create an SQLiteConnection in a Windows 8.1 project, it's code that works fine on Android and iOS builds. The steps to reproduce it are as follows:

1. Create a Blank Xaml App (Xamarin.Forms.Portable) in Visual Studio Community
2. Use NuGet manager to add sqlite-net-pcl v1.3.1 to the Windows 8.1 project
3. Add the following function to the Windows 8.1 App class:

```
protected override void OnWindowCreated(WindowCreatedEventArgs args)
{
	base.OnWindowCreated(args);

	try
	{
		var path = ApplicationData.Current.LocalFolder.Path + @"\mydb.db3";
		var connection = new SQLiteConnection(path, SQLiteOpenFlags.Create);
		connection.Close();
	}
	catch (Exception e)
	{
		Debug.WriteLine(e.Message);
		if (e.InnerException != null)
			Debug.WriteLine(e.InnerException.Message);
	}
}
```

4. Build and run the Windows 8.1 application. The exception is thrown and the following output generated:
```
The type initializer for 'SQLite.SQLiteConnection' threw an exception.
Unable to load DLL 'e_sqlite3': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
```

Is this a bug, or do I need to do something different on Windows builds?

**Comment by ericsink** [comment]
Did you add the extension SDK for the Visual C++ runtime?

**Comment by Myndale** [comment]
Thanks for the quick response Eric. And no, I only followed the instructions in the "Nuget Installation" section in the README.md. Sorry if I'm missing something obvious here but I can't see any reference to any "extension SDK", is there a NuGet name and/or URL you could point me towards? Thanks for the help!

**Comment by ericsink** [comment]
From memory, not sure I've got this right, or even that it will fix the problem.

It's not a nuget package.

In visual studio, add reference.  Choose Extensions.  Look for the visual c++ runtime in there.

**Comment by Myndale** [comment]
Tried that, still no-go unfortunately :( I've downloaded the sqlite-net project and linked my project directly to it, the exception's being thrown almost immediately by the SQLite3.SetDirectory call, and SQLite3 is being aliased to your SQLitePCL.raw library. I guess maybe the next step is to download SQLitePCL.raw and try to step through that?

**Comment by ericsink** [comment]
I would stick with the nuget package. 

When I get back to a computer and have a chance, I will try your repro steps.

**Comment by Myndale** [comment]
Thanks Eric, appreciate it. I do plan to stick to NuGet, just downloaded the project to try and track this issue down.

**Comment by ericsink** [comment]
I reproduced this problem, and adding the Visual C++ 2013 runtime library extension SDK did resolve it for me.

**Comment by Myndale** [comment]
Which package library does your reference resolve to? I've gone and looked at my project again and the resolve is failing, even though the actual folder (i.e. C:\Program Files (x86)\Microsoft SDKs\Windows\v8.1\ExtensionSDKs\Microsoft.VCLibs\12.0) is present on my drive. The only libs I see though are for arm, x86 and x64....I'm guessing there should be one for portable or win8 or something?

**Comment by ericsink** [comment]
I can't tell what you are referring to.

Are you getting any error about building with Any CPU?  When that extension SDK is added, you need to build for a specific CPU. 

Just to be sure:

Add Reference

In the Reference Manager dialog, under Windows 8.1, click on Extensions

You should see "Microsoft Visual C++ 2013 Runtime Package for Windows".  Version 12.0.  Choose that.

Then switch your build to x64 instead of Any CPU.

The steps for Windows Phone 8.1 and Windows 8.1 are rather tedious.  :-)

**Comment by apspot** [comment]
Hello,

I'm getting the same error :( Please help me because I'm unable to solve this.

![image](https://cloud.githubusercontent.com/assets/7696301/25675496/ef89ced8-303e-11e7-8a2a-3431945fdd06.png)

![image](https://cloud.githubusercontent.com/assets/7696301/25675578/47b656ee-303f-11e7-873b-0e301df766c6.png)

![image](https://cloud.githubusercontent.com/assets/7696301/25675626/76544182-303f-11e7-9759-2c43e1c06628.png)

**Comment by ericsink** [comment]
You shouldn't need "SQLite for Windows Runtime (Windows 8.1)", but that won't solve your problem.

Your other settings look fine, so I don't immediately see what is wrong.

You should first check to see if the e_sqlite3.dll file is in fact appearing somewhere in your build output directory.

**Comment by apspot** [comment]
There is no e_sqlite3.dll in the output folder, but I have SQLitePCLRaw.provider.e_sqlite3.dll. It was installed as a dependency. Do you know what is this "e_sqlite3.dll"? Can I add it via NuGet? I tried to Google it but I found nothing useful :(

The Android project doesn't have e_sqlite3.dll in the output but it works perfectly on my phone.

Thank you very much for your help!

**Comment by ericsink** [comment]
e_sqlite3.dll contains the native code for SQLite.  It should be provided to your project by the bundle_green package which is a dependency of sqlite-net-pcl.

Is this Windows 8.1 silverlight, or the RT flavor?  I'm guessing the RT flavor, based on information in your screen shot.

If by chance I could see your csproj file, it might help me figure out what's going on.

**Comment by apspot** [comment]
Here it is: https://github.com/apspot/HelloXamarin/blob/master/HelloXamarin/HelloXamarin.Windows/HelloXamarin.Windows.csproj

I created this demo project while I was learning Xamarin Forms on an online course.

Thank you for your help again :)

**Comment by praeclarum** [comment]
Hello,

Have you tried upgrading to the v1.4? I'm curious if this is still a problem.

Thanks!

**Comment by Myndale** [comment]
Hi Frank,

 

Haven’t been able to test it I’m afraid, that issue was for a client that I did some contract work for earlier on this year. To be honest I think they eventually dropped the Win8 build altogether because of the many problems we encountered with the Xamarin framework itself. If I do get a chance to work on a Win8 project again in future I’ll definitely give it another go, sqlite-net was terrific to work with on the other platforms.

 

Cheers,

 

Mark
